core_backend/app/llm_call/parse_input.py

```diff
@@ -2,75 +2,196 @@
+    Decorator to identify the language of the question.
     """
-    if classify_safety(question.query_text) == SafetyClassification.SAFE:
-        return True
-    else:
-        return False
-
 
-def identify_language(question: UserQueryRefined) -> UserQueryRefined:
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to identify the language of the question.
+        """
+        question, response = _identify_language(question, response)
+        response = await func(question, response, *args, **kwargs)
+        return response
+
+    return wrapper
+
+
+def _identify_language(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Identifies the language of the question.
     """
+    if response.state != ResultState.ERROR:
+        question.original_language = getattr(
+            IdentifiedLanguage,
+            _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
+        )
+        if question.original_language is not None:
+            response.debug_info["original_language"] = question.original_language.value
 
-    question.original_language = getattr(
-        IdentifiedLanguage,
-        _ask_llm(question.query_text, IdentifiedLanguage.get_prompt()),
-    )
+    return question, response
 
-    return question
 
+def translate_question(func: Callable) -> Callable:
+    """
+    Decorator to translate the question.
+    """
 
-def translate_question(question: UserQueryRefined) -> UserQueryRefined:
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to translate the question.
+        """
+        question, response = _identify_language(question, response)
+        question, response = _translate_question(question, response)
+        response = await func(question, response, *args, **kwargs)
+        response.debug_info["translated_question"] = question.query_text
+
+        return response
+
+    return wrapper
+
+
+def _translate_question(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Translates the question to English.
     """
 
     if (
-        question.original_language
-        in [
-            IdentifiedLanguage.ENGLISH,
-            IdentifiedLanguage.UNKNOWN,
-        ]
-        or question.original_language is None
+        question.original_language == IdentifiedLanguage.ENGLISH
+        or response.state == ResultState.ERROR
     ):
-        return question
+        return question, response
+
+    if question.original_language is None:
```

Comment:
Will never happen given how the code is right now since we run identify language everytime we run translate. But is a safe programming to assume as little as possible about what happens outside of the function.